File: meat-trade/chaincode/slaughterhouse/node/slaughterhouse.js
/*******************************************************************************
 * Copyright 2018 CPqD. All Rights Reserved.
 * 
 * SPDX-License-Identifier: Apache-2.0
 */

const shim = require('fabric-shim')

const SUPERMARKET_COLLECTION = 'collectionSupermarket1PD';

const ALLOWED_MSPS_REGISTER = ['SlaughterhouseMSP'];
const ALLOWED_MSPS_QUERY = ['SlaughterhouseMSP', 'Inspection1MSP', 'Inspection2MSP'];

var SHouseChaincode = class {

  // Initialize the chaincode
  async Init(stub) {    
    return shim.success("SHouseChaincode successfully instantiated");
  }

  async Invoke(stub) {
    let ret = stub.getFunctionAndParameters();
    console.info(ret);
    let method = this[ret.fcn];
    if (!method) {
      console.log('method with name :' + ret.fcn + ' does not exist');
      shim.error(new Error('method with name :' + ret.fcn + ' does not exist'));
    }
    try {
      let payload = await method(stub, ret.params);
      return shim.success(payload);
    } catch (err) {
      console.log(err);
      return shim.error(err);
    }
  }

  /**
   * Registers public sale on the ledger
   * @param {ChaincodeStub} stub 
   * @param {Array} args 
   */
  async registerSale(stub, args) {

    if (args.length < 2) {
      throw new Error('Wrong number of arguments, please provide 2.');
    }

    let creator = stub.getCreator();
    if (!creator) {
      throw new Error('Failed getting request creator');
    }

    if (ALLOWED_MSPS_REGISTER.indexOf(creator.mspid) == -1) {
      throw new Error('Organization not authorized to perform this action.');
    }

    let invoiceNumber = args[0];    
    let sale = args[1];

    let queryTransaction = await stub.getState(invoiceNumber);
    if (queryTransaction && queryTransaction.toString().length > 0) {
      throw new Error('Unable to perform action. Key already exists.');
    }


    try {
      var carcases = JSON.parse(sale);
    } catch (convertException) {
      console.log('JSON parse error');
      throw new Error('JSON parse error');
    }

    var transaction = carcases;

    await stub.putState(invoiceNumber, Buffer.from(JSON.stringify(transaction)));

  }

  /**
   * Queries public sale on ledger
   * @param {ChaincodeStub} stub 
   * @param {Array} args 
   */
  async querySale(stub, args) {
    let invoiceNumber = args[0];
    if (!invoiceNumber) {
      throw new Error('Wrong number of arguments, please provide 1.');
    }

    let transactionAsBytes = await stub.getState(invoiceNumber);
    if (!transactionAsBytes || transactionAsBytes.toString().length <= 0) {
      console.log('Transaction not found');
      throw new Error('Transaction not found');
    }
    return transactionAsBytes;
  }


  /**
   * Registers private sale on ledger.
   * @param {ChaincodeStub} stub
   * @param {Array} args 
   */
  async registerPrivateSale(stub, args) {
    if (args.length < 2) {
      throw new Error('Wrong number of arguments, please provide 2.');
    }

    let invoiceNumber = args[0];    
    let sale = args[1];

    let queryTransaction = await stub.getState(invoiceNumber);
    if (queryTransaction && queryTransaction.toString().length > 0) {
      throw new Error('Unable to perform action. Key already exists.');
    }


    try {
      var carcases = JSON.parse(sale);
    } catch (convertException) {
      console.log('JSON parse error');
      throw new Error('JSON parse error');
    }

    var transaction = carcases;
    
    let creator = stub.getCreator();
    if (!creator) {
      throw new Error('Failed getting request creator');
    }

    if (ALLOWED_MSPS_REGISTER.indexOf(creator.mspid) == -1) {
      throw new Error('Organization not authorized to perform this action.');
    }
    await stub.putPrivateData(SUPERMARKET_COLLECTION, invoiceNumber, Buffer.from(JSON.stringify(transaction)));
    
  }

  /**
   * Queries private sale on ledger
   * @param {ChaincodeStub} stub 
   * @param {Array} args 
   */
  async queryPrivateSale(stub, args) {
    let invoiceNumber = args[0];
    if (!invoiceNumber) {
      throw new Error('Wrong number of arguments, please provide 1.');
    }
    
    let creator = stub.getCreator();
    if (!creator) {
      throw new Error('Failed getting request creator');
    }

    if (ALLOWED_MSPS_QUERY.indexOf(creator.mspid) == -1) {
      throw new Error('Organization not authorized to perform this action.');
    }

    let transactionAsBytes = await stub.getPrivateData(SUPERMARKET_COLLECTION, invoiceNumber);
    if (!transactionAsBytes || transactionAsBytes.toString().length <= 0) {
      console.log('Transaction not found');
      throw new Error('Transaction not found');
    }
    return transactionAsBytes;
  }

  /**
   * Get history for a given on the ledger
   * @param {ChaincodeStub} stub 
   * @param {Array} args 
   */
  async getHistoryForKey(stub, args) {

    let key = args[0];
    if (!key) {
      throw new Error('Wrong number of arguments, please provide 1');
    }

    let iterator = await stub.getHistoryForKey(key);

    let allResults = [];

    while (true) {
      let res = await iterator.next();
      if (res.value && res.value.value.toString()) {
        let jsonRes = {};
        jsonRes.key = res.value.key;
        jsonRes.isDeleted = res.value.is_delete;
        jsonRes.transactionId = res.value.tx_id;
        try {
          jsonRes.record = JSON.parse(res.value.value.toString('utf8'));
        } catch (err) {
          console.log(err);
          jsonRes.record = res.value.value.toString('utf8');
        }
        allResults.push(jsonRes);
      }

      if (res.done) {
        console.log('end of data');
        await iterator.close();
        console.info(allResults);
        return Buffer.from(JSON.stringify(allResults));
      }
    }
  }
};

let chaincode = new SHouseChaincode();

shim.start(chaincode);

File: meat-trade/chaincode/slaughterhouse/go/slaughterhouse.go
/*******************************************************************************
 * Copyright 2018 CPqD. All Rights Reserved.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"strconv"

	"github.com/gogo/protobuf/proto"
	"github.com/hyperledger/fabric/core/chaincode/shim"
	"github.com/hyperledger/fabric/protos/msp"
	sc "github.com/hyperledger/fabric/protos/peer"
)

/** Collections **/

//Supermarket1Collection private data
const Supermarket1Collection string = "collectionSupermarket1PD"

//MspSupermarket1 MSP of supermarket
const MspSupermarket1 string = "Supermarket1MSP"

//MspInspection2 MSP of inspection2 org
const MspInspection2 string = "Inspection2MSP"

//MspInspection1 MSP of inspection1 org
const MspInspection1 string = "Inspection1MSP"

//MspSlaughterhouse MSP of slaughterhouse org
const MspSlaughterhouse string = "SlaughterhouseMSP"

//NotAllowedOrg error message
const NotAllowedOrg string = "Organization not authorized to perform this action."

//WrongNumberOfArgs error message
const WrongNumberOfArgs string = "Wrong number of arguments, please provide %v."

//ErrorMessagePermissionDenied Error Message
const ErrorMessagePermissionDenied string = "Organization not authorized to perform this action."

//CutsOfBeef structure
type CutsOfBeef struct {
	Slaughterhouse string `json:"slaughterhouse"`
	Market         string `json:"market"`
	Carcases       []struct {
		IDAnimal string `json:"idAnimal"`
		ID       string `json:"id"`
		Weight   string `json:"weight"`
		Type     string `json:"type"`
	} `json:"carcases"`
}

//Init method used to initialize blockchain
func (p *CutsOfBeef) Init(APIstub shim.ChaincodeStubInterface) sc.Response {
	return shim.Success(nil)
}

//Invoke methods
func (p *CutsOfBeef) Invoke(APIstub shim.ChaincodeStubInterface) sc.Response {
	function, args := APIstub.GetFunctionAndParameters()
	// Route to the appropriate handler function to interact with the ledger appropriately
	if function == "registerSale" {
		return p.registerSale(APIstub, args)
	} else if function == "querySale" {
		return p.querySale(APIstub, args)
	} else if function == "getHistoryForKey" {
		return p.getHistoryForKey(APIstub, args)
	} else if function == "registerPrivateSale" {
		return p.registerPrivateSale(APIstub, args)
	} else if function == "queryPrivateSale" {
		return p.queryPrivateSale(APIstub, args)
	}

	return shim.Error("method with name :" + function + " does not exist")
}

func (p *CutsOfBeef) registerSale(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	if len(args) < 2 {
		return shim.Error(fmt.Sprintf(WrongNumberOfArgs, 2))
	}

	danfe := args[0]
	bytes := []byte(args[1])

	var beefCatle CutsOfBeef

	if err := json.Unmarshal(bytes, &beefCatle); err != nil {
		return shim.Error(err.Error())
	}

	beefCatlesAsByte, err := json.Marshal(beefCatle)

	if err != nil {
		return shim.Error(err.Error())
	}
	mspRequestID, err := p.getMspid(APIstub)

	if mspRequestID == MspSlaughterhouse {
		if err := APIstub.PutState(danfe, beefCatlesAsByte); err != nil {
			return shim.Error(err.Error())
		}
	} else {
		return shim.Error(NotAllowedOrg)
	}

	return shim.Success(nil)
}

func (p *CutsOfBeef) querySale(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	if len(args) < 1 {
		return shim.Error(fmt.Sprintf(WrongNumberOfArgs, 1))
	}

	danfe := args[0]

	beefCatlesAsByte, err := APIstub.GetState(danfe)

	if err != nil {
		return shim.Error(err.Error())
	}

	return shim.Success(beefCatlesAsByte)
}

func (p *CutsOfBeef) getHistoryForKey(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	if len(args) != 1 {
		return shim.Error(fmt.Sprintf(WrongNumberOfArgs, 1))
	}

	danfe := args[0]

	historyIterator, err := APIstub.GetHistoryForKey(danfe)

	if err != nil {
		return shim.Error(err.Error())
	}
	defer historyIterator.Close()

	var buffer bytes.Buffer
	buffer.WriteString("[")

	bArrayMemberAlreadyWritten := false

	for historyIterator.HasNext() {
		keyModification, err := historyIterator.Next()

		if err != nil {
			return shim.Error(err.Error())
		}

		if bArrayMemberAlreadyWritten {
			buffer.WriteString(",")
		}
		json.Marshal(keyModification)
		buffer.WriteString("{\"TxId\":")
		buffer.WriteString("\"")
		buffer.WriteString(keyModification.TxId)
		buffer.WriteString("\"")

		buffer.WriteString(", \"Value\":")
		// Record is a JSON object, so we write as-is
		buffer.WriteString(string(keyModification.Value))

		buffer.WriteString(", \"IsDelete\":")
		// Record is a JSON object, so we write as-is
		buffer.WriteString(strconv.FormatBool(keyModification.IsDelete))
		buffer.WriteString("}")

		bArrayMemberAlreadyWritten = true
	}

	buffer.WriteString("]")
	return shim.Success(buffer.Bytes())
}

func (p *CutsOfBeef) registerPrivateSale(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	if len(args) < 2 {
		return shim.Error(fmt.Sprintf(WrongNumberOfArgs, 2))
	}

	danfe := args[0]
	bytes := []byte(args[1])

	mspid, err := p.getMspid(APIstub)

	if err != nil {
		return shim.Error(err.Error())
	}

	if mspid != MspSlaughterhouse {
		return shim.Error(ErrorMessagePermissionDenied)
	}

	var beefCatle CutsOfBeef

	if err := json.Unmarshal(bytes, &beefCatle); err != nil {
		return shim.Error(err.Error())
	}

	beefCatlesAsByte, err := json.Marshal(beefCatle)

	if err != nil {
		return shim.Error(err.Error())
	}

	if err := APIstub.PutPrivateData(Supermarket1Collection, danfe, beefCatlesAsByte); err != nil {
		return shim.Error(err.Error())
	}

	return shim.Success(nil)
}

func (p *CutsOfBeef) queryPrivateSale(APIstub shim.ChaincodeStubInterface, args []string) sc.Response {
	if len(args) < 1 {
		return shim.Error(fmt.Sprintf(WrongNumberOfArgs, 1))
	}

	danfe := args[0]

	mspid, err := p.getMspid(APIstub)

	if err != nil {
		return shim.Error(err.Error())
	}

	collection, err := p.getCollectionNameByMspid(mspid)

	if err != nil {
		return shim.Error(err.Error())
	}

	beefCatlesAsByte, err := APIstub.GetPrivateData(collection, danfe)

	if err != nil {
		return shim.Error(err.Error())
	}

	return shim.Success(beefCatlesAsByte)
}

func (p *CutsOfBeef) getMspid(APIstub shim.ChaincodeStubInterface) (string, error) {
	serializedCreator, err := APIstub.GetCreator()

	if err != nil {
		return "", errors.New(err.Error())
	}

	creator := &msp.SerializedIdentity{}

	if err := proto.Unmarshal(serializedCreator, creator); err != nil {
		return "", errors.New(err.Error())
	}

	return creator.Mspid, nil
}

func (p *CutsOfBeef) getCollectionName(APIstub shim.ChaincodeStubInterface) (string, error) {
	mspid, err := p.getMspid(APIstub)

	if err != nil {
		return "", errors.New(err.Error())
	}

	return p.getCollectionNameByMspid(mspid)
}

func (p *CutsOfBeef) getCollectionNameByMspid(Mspid string) (string, error) {

	if Mspid == MspSlaughterhouse || Mspid == MspSupermarket1 || Mspid == MspInspection1 || Mspid == MspInspection2 {
		return Supermarket1Collection, nil
	} else {
		return "", errors.New(ErrorMessagePermissionDenied)
	}
}

func main() {

	// Create a new Smart Contract
	err := shim.Start(new(CutsOfBeef))
	if err != nil {
		fmt.Printf("Error creating new Smart Contract: %s", err)
	}
}
